driver-core/src/main/com/mongodb/internal/connection/IamAuthenticator.java

```diff
@@ -0,0 +1,275 @@
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public byte[] wrap(final byte[] bytes, final int i, final int i1) throws SaslException {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public Object getNegotiatedProperty(final String s) {
+            throw new UnsupportedOperationException("Not implemented yet!");
+        }
+
+        @Override
+        public void dispose() throws SaslException {
+            // nothing to do
+        }
+
+        private byte[] computeClientFirstMessage() throws SaslException {
+            new SecureRandom(this.clientNonce).nextBytes(this.clientNonce);
+
+            BsonDocument document = new BsonDocument()
+                    .append("r", new BsonBinary(this.clientNonce))
+                    .append("p", new BsonInt32('n'));
+            return toBson(document);
+        }
+
+        private byte[] computeClientFinalMessage(final byte[] serverFirst) throws SaslException {
+            final BsonDocument document = new RawBsonDocument(serverFirst);
+            final String host = document.getString("h").getValue();
+
+            final byte[] serverNonce = document.getBinary("s").getData();
+            if (serverNonce.length != (2 * RANDOM_LENGTH) || !Arrays.equals(Arrays.copyOf(serverNonce, RANDOM_LENGTH), this.clientNonce)) {
+                throw new SaslException("Invalid server nonce");
+            }
+
+            String timestamp = DateTimeFormatter.ofPattern("yyyyMMdd'T'HHmmss'Z'")
+                    .withZone(ZoneId.of("UTC"))
+                    .format(Instant.now());
+
+            String token = getSessionToken();
+            final AuthorizationHeader authorizationHeader = AuthorizationHeader.builder()
+                    .setAccessKeyID(getUserName())
+                    .setSecretKey(getPassword())
+                    .setSessionToken(token)
+                    .setHost(host)
+                    .setNonce(serverNonce)
+                    .setTimestamp(timestamp)
+                    .build();
+
+            BsonDocument ret = new BsonDocument()
+                    .append("a", new BsonString(authorizationHeader.toString()))
+                    .append("d", new BsonString(authorizationHeader.getTimestamp()));
+            if (token != null) {
+                ret.append("t", new BsonString(token));
+            }
+
+            return toBson(ret);
+        }
+
+
+        private byte[] toBson(final BsonDocument document) {
+            byte[] bytes;
+            BasicOutputBuffer buffer = new BasicOutputBuffer();
+            new BsonDocumentCodec().encode(new BsonBinaryWriter(buffer), document, EncoderContext.builder().build());
+            bytes = new byte[buffer.size()];
+            int curPos = 0;
+            for (ByteBuf cur : buffer.getByteBuffers()) {
```

Comment:
I think you can just do:

   System.arraycopy(buffer.getInternalBuffer(), 0, bytes, 0, buffer.getSize()) 

and get rid of the loop

Reply:
Done